internal/consensus/state.go

```diff
@@ -2070,20 +2057,16 @@ func (cs *State) addVote(vote *types.Vote, peerID types.NodeID) (added bool, err
 		prevotes := cs.Votes.Prevotes(vote.Round)
 		cs.Logger.Debug("added vote to prevote", "vote", vote, "prevotes", prevotes.StringShort())
 
-		// If +2/3 prevotes for a block or nil for *any* round:
+		// If +2/3 prevotes for a block for *any* round:
 		if blockID, ok := prevotes.TwoThirdsMajority(); ok {
 			// There was a polka!
-			// If we're locked but this is a recent polka, unlock.
 			// If it matches our ProposalBlock, update the ValidBlock
-
-			// Unlock if `cs.LockedRound < vote.Round <= cs.Round`
-			// NOTE: If vote.Round > cs.Round, we'll deal with it when we get to vote.Round
 			if (cs.LockedBlock != nil) &&
 				(cs.LockedRound < vote.Round) &&
 				(vote.Round <= cs.Round) &&
-				!cs.LockedBlock.HashesTo(blockID.Hash) {
+				!cs.LockedBlock.HashesTo(blockID.Hash) && len(blockID.Hash) != 0 {
 
-				cs.Logger.Debug("unlocking because of POL", "locked_round", cs.LockedRound, "pol_round", vote.Round)
+				cs.Logger.Debug("unlocking because of POL for non-nil block", "locked_round", cs.LockedRound, "pol_round", vote.Round)
 
 				cs.LockedRound = -1
 				cs.LockedBlock = nil
```

Comment:
So we have 2/3+ prevotes for a block that's not the one we locked on and so we unlock. Do we lock on this new block in the same round and precommit it? Or precommit nil. It's not clear what action we are doing at this stage. On line 1432 it says to stage and precommit it but I don't see where we do that

Reply:
From my investigation, the unlock here is entirely useless. I'm planning on cleaning it up in #6850

If you notice the update to the TestStateLockPOLRelock, you can see we check for an unlock followed by a lock within the same consensus round. 
https://github.com/tendermint/tendermint/blob/c69f933d61962ffae5420b1cb476bfa2e5c96a1f/internal/consensus/state_test.go#L704

This means that the unlock step here is not doing anything since we're going to relock within the same round. It's therefore totally safe to remove.

Reply:
Agree this is not needed. Maybe it have been better removing unlock logic in this PR in general, not just in case where there is 2/3+ nil prevotes.